### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by 522122** [hhd-dev/hhd-ui#13]
Noticed server responses are kind of slow. Polling takes 2 sec and applying setting around 1 sec. This feels sometimes slow, especially when updating the sliders with controller. Every step you get 1 sec "lag"

<img width="1349" alt="Screenshot 2024-08-13 at 11 55 42" src="https://github.com/user-attachments/assets/295028cb-9f91-4794-88d0-791ff0fca99b">

**Comment by antheas** [comment]
Hi. 

Yes, handheld daemon has a polling endpoint that returns when its inner loop runs. This is either every 2 seconds or when a setting changes.

You can verify that when messing with the ppd slider in KDE. It updates instantly.

Certain settings are either debounced or have a small sleep in them for the purposes of system stability.

I made sure you can barely notice.

As for what you are noticing. Unfortunately one of the architectural pain points I made was that I never debounced the controller settings. This means that every time you change a controller setting, the controller reloads. 

This happens within 200ms. However, you get an axis slip if you are currently modifying a setting (you have to release the stick for it to work again). 

In addition, if you are using the Gamepad API for controller support in the UI (happens when using the desktop shortcut) it will take around a second to reregister the controller.

This is what I think you are noticing. As this is one of the very few parts in Handheld Daemon that is duplicated per manufacturer, it is a bit difficult to change now.

**Comment by antheas** [comment]
As for applying settings, most TDP settings are debounced and apply after 2 seconds purposely. This is to avoid overloading the embedded computer of the device.

**Comment by 522122** [comment]
Thanks for explaining.
Sorry, don't know what `Gamepad API` means. I get the same 1s delay when using Ally's controller or PS5 controller connected to my mac and doing it on dev in edge browser.

And by testing it again it seems to happen only for charging limit. Other settings are much quicker to respond.

On my screenshot there is 1 request that is not poll that took 1s and that was moving the charge limit slider.

**Comment by antheas** [comment]
Here is a reference for gamepad API.
https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API/Using_the_Gamepad_API

I tested the Asus Slider. On the stable build, it works instantly. Therefore, it is either a kernel bug that hangs handheld daemon while it is writing it (I do not have an ally in hand to test this) or it is a bug in your branch.

If it is a kernel bug there is not much that can be done to fix this, as the python interpreter will freeze while writing the value (the code could be refactored to only work after apply is hit; where it would not be noticeable).
In this case, the best place to report this is here: https://gitlab.com/asus-linux/asusctl

If it is a problem with your PR, then happy testing.

**Comment by 522122** [comment]
Indeed writing to `/sys/class/power_supply/BAT0/charge_control_end_threshold` takes exactly 1 sec on Ally X. Should it be faster?

**Comment by antheas** [comment]
Yes, it should use a worker thread and be instant. This can't be fixed in software as the interpreter lock is not released when writing to files. 

Which means that the whole handheld daemon process freezes while changing the charging limit.

We can debounce it. So let me transfer this issue. However, don't expect me to fix it any time soon.

I'll probably do it in a month or so
